### DISCUSSION TITLE
Data from different units

### DISCUSSION THREAD

**Discussion by muxa** [issalig/toshiba_air_cond#18] [General]
I'm building the interface for my parent's air con:

* Remote: `RBC-AMS41E`
* AC: `MMD-AP0271BH`

I'm capturing the data frames for different commands etc. Mostly it's the same as @issalig captured, but there are some differences. I'm starting this discussion for people to share their data capture so we can better understand the differences and hopefully learn more about the TCC-Link protocol.

* The most obvious difference is instead of `00` for master I have `01`. I'm guessing this is to support multiple units from a single remote, and in my case the unit is configured as `01`.  E.g:
  * ALIVE message: `01:FE:10:02:80:8A:E7 `
  * ACK after setting a parameter: `01:40:18:02:80:A1:7A` 
  * PING: `40:01:15:07:08:0C:81:00:00:48:00:9E`
  * PONG: `01:40:18:08:80:0C:82:03:00:00:48:00:14`
* Status message is 5 bytes shorter, e.g: `01:FE:1C:08:80:81:35:6C:02:00:78:08:C1`
* Extended status message is 4 bytes shorter, e.g: `01:FE:58:0A:80:81:35:6C:02:00:78:67:E9:00:01`

**Comment by issalig** [comment]
Congratulations! You are the first one that gives feedback about communication.
Maybe I should put master "id" as a configurable field. As far as I remember messages are hardcoded and they just use "00"

Let me know if you discover more things.

**Comment by muxa** [comment]
Not sure if different to your unit, but more details for my use case:

```
01:52:11:04:80:86:A1:05:E4
                  |  |
                  |  |- & 0b00000001: POWER (ON/OFF)
                  |  |- & 0b00000100 >> 2: HEATING 1 (ON/OFF)
                  |
                  |- & 0b11100000 >> 5: MODE
                  |- & 0b00001000 >> 3: COOLING (ON/OFF)
                  |- & 0b00000001: HEATING 2? (ON/OFF)

```

There seem to be 2 flags that correspond to HEATING status. They are not always the same.  E.g. in DRY mode HEATING 1 is `1` and HEATING 2 is `0`. I'll observe this for a bit longer.

**Comment by muxa** [comment]
Also some differences for setting temp or fan. 

For setting _fan_ the mode byte is `0b00010000 | mode`. 
For setting _temperature_ the mode byte is `0b00001000 | mode`. 

The fan byte is set to `0b00101000 | fan` when heating and `0b00011000 | fan` when cooling

Here's my implementation snippet of the write logic:

```cpp
uint8_t get_fan_bit_mask_for_mode(uint8_t mode) {
  switch (mode) {
    case MODE_HEAT:
    case MODE_AUTO:
      return 0b00100000 | 0b00001000;
    case MODE_COOL:
    case MODE_DRY:
    case MODE_FAN_ONLY:
      return 0b00010000 | 0b00001000;
  }

  return 0;
}

void write_set_parameter_flags(struct DataFrame *data_frame, struct StateStore *state_store, uint8_t set_flags) {
  uint8_t payload[6] = {
    static_cast<uint8_t>(state_store->mode | set_flags),
    static_cast<uint8_t>(state_store->fan | get_fan_bit_mask_for_mode(state_store->mode)),
    temp_celcius_to_payload(state_store->target_temp),
    EMPTY_DATA,
    get_heat_cool_bits(state_store->mode), 
    get_heat_cool_bits(state_store->mode), 
  };
  write_set_parameter(data_frame, state_store, 0x4C, payload, sizeof(payload));
}

// the flags
const uint8_t COMMAND_SET_TEMP = 0b00001000;
const uint8_t COMMAND_SET_FAN = 0b00010000;
```

**Reply by muxa** [reply]
I have published my current implementation here: https://github.com/muxa/esphome-tcc-link
